alerting.md

```diff
@@ -0,0 +1,109 @@
+
+### Monitoring the alert
+
+After the alert has been put to production you have ownership of it. This means you have to track its accountability. When it’s raised you should answer these questions:
+Is the alert catching the symptom?
+Was the alert raised in timely fashion?
+Was it a genuine alert (signal) and not a fake one (noise)?
+Did you take any actions in order to fix it?
+
+	If all above questions are answered affirmatively, congratulations! If not, iterate!
+
+### New Relic (NR)
+
+Pros:
+* Setting up response times and errors per user action (i.e. high error rate on user registration)
+* Pinging an application (pingdom-like health check, checks if the application is online and responds with success)
+* Tracking response times on the web server
+
+Cons:
+* Very sensitive when monitoring end user response times (slow clients)
+* Possible false alerts when low traffic (i.e. 3 requests in last 5 minutes, ⅔ were errors => 66% error rate)
+
+### DataDog (DD)
+
+Pros:
+* Background process healthcheck (i.e. scheduler is working)
+* Ensuring background jobs are being processed (i.e. properties being imported, rates synced)
+* Define custom alert required for the business (i.e. queue staleness alerts)
+
+Cons:
+* The logic that can be applied to a monitored action is pretty rudimentary. You can only set alerting rules on a single monitored action.
+* Monitoring low volume actions (couple of events per day) makes it hard to define reactive alerts
+
+### PagerDuty (PD)
+
+PagerDuty is a tool that connects with external services like DD and NR. It’s main responsibility is ensuring the people who are on duty receive alerts.
+It also allows us to view a report of past alerts and write notes on an individual alert.
+
+## What to do when I get alerted?
+
+* Don’t panic.
+* Read the alert message and acknowledge the alert on PagerDuty. It’s important to acknowledge the alert so it doesn’t escalate and send email to every engineer in roots@housetrip.com group, or SMS messages to non duty-engineers.
+* The alert message tells you about the symptom, your mission is to identify the cause. The message should point you in the right direction. i.e. “no enquiry email sent” should point towards the DJ not being processed.
+* Fixing the issue. Don’t spend too much time on a perfect solution but find one that does the job and fixes the cause. Usually:
+* turn off a feature flag
+* temporary disable an ab test
+* disable the feature with a hotfix
+* Write a note on the PagerDuty alert. Be brief, don’t spend more time on documenting the alert than it took you fixing it. The note should include:
+* What was the cause? Bonus points for including links to useful dashboards on NR or DD
+* How did you fix the cause; useful for future duty comrades
+* If the alert was not real or unactionable, remember to point that out. We need this info to continuously improve and maintain a good alert set. Again fellow engineers will appreciate it.
+
+## Caveats
```

Comment:
Suggested changed for this last section:

```
# Caveats

All this symptom based alerting is nice and brings the mindset closer to the end user’s, but what does one do when we have to deal with special situations like:

### Detecting very rare failures (< 1%)

In this case, setting up the alert on symptoms would generate noise or ignore the cause. You don’t want to create noisy alerts!

### Symptom based alerting comes too late

An example here is how to deal with running out of quotas (free disk space, free Redis memory, …). Instead of alerting too late (less than a couple of hours until running out of storage), send a notification which raises the issue earlier (if we continue to grow at this rate, we will run out of quota in 1+ days) that can be resolved in office hours.

### Symptom-based alerting is too complicated

This goes hand in hand with the previous point. On a few rare occasions defining the symptom becomes overwhelming, hard to maintain and understand (i.e. read replica has become inaccessible). If it’s much simpler and accurate to alert on cause compared to the symptom, it’s worth making an exception to the rule.
```